pstoreds/addr_book.go

```diff
@@ -2,398 +2,608 @@ package pstoreds
+func (r *addrsRecord) FlushInTxn(txn ds.Txn) (err error) {
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(r.Id.ID)))
+	if len(r.Addrs) == 0 {
+		return txn.Delete(key)
+	}
+	// cannot use a buffer pool because data is retained in the txn until it's committed or discarded.
+	data, err := r.Marshal()
+	if err != nil {
+		return err
+	}
+	if err = txn.Put(key, data); err != nil {
+		return err
+	}
+	// write succeeded; record is no longer dirty.
+	r.dirty = false
 	return nil
 }
 
-// NewAddrBook initializes a new address book given a
-// Datastore instance, a context for managing the TTL manager,
-// and the interval at which the TTL manager should sweep the Datastore.
-func NewAddrBook(ctx context.Context, store ds.TxnDatastore, opts Options) (*dsAddrBook, error) {
-	if _, ok := store.(ds.TTLDatastore); !ok {
-		return nil, ErrTTLDatastore
+// Flush creates a ds.Txn, and calls FlushInTxn with it.
+func (r *addrsRecord) Flush(ds ds.TxnDatastore) (err error) {
+	txn, err := ds.NewTransaction(false)
+	if err != nil {
+		return err
 	}
+	defer txn.Discard()
 
-	var (
-		cache cache = &noopCache{}
-		err   error
-	)
+	if err = r.FlushInTxn(txn); err != nil {
+		return err
+	}
+	return txn.Commit()
+}
 
-	if opts.CacheSize > 0 {
-		if cache, err = lru.NewARC(int(opts.CacheSize)); err != nil {
-			return nil, err
+// Refresh is called on records to perform housekeeping. The return value signals if the record was changed
+// as a result of the refresh.
+//
+// Refresh does the following:
+// * sorts the addresses by expiration (soonest expiring first).
+// * removes the addresses that have expired.
+//
+// It short-circuits optimistically when we know there's nothing to do.
+//
+// Refresh is called from several points:
+// * when accessing and loading an entry.
+// * when performing periodic GC.
+// * after an entry has been modified (e.g. addresses have been added or removed,
+//   TTLs updated, etc.)
+//
+// If the return value is true, the caller can perform a flush immediately, or can schedule an async
+// flush, depending on the context.
+func (r *addrsRecord) Refresh() (chgd bool) {
+	now := time.Now().Unix()
+	if !r.dirty && len(r.Addrs) > 0 && r.Addrs[0].Expiry > now {
+		// record is not dirty, and we have no expired entries to purge.
+		return false
+	}
+
+	if len(r.Addrs) == 0 {
+		// this is a ghost record; let's signal it has to be written.
+		// Flush() will take care of doing the deletion.
+		return true
+	}
+
+	if r.dirty && len(r.Addrs) > 1 {
+		// the record has been modified, so it may need resorting.
+		// we keep addresses sorted by expiration, where 0 is the soonest expiring.
+		sort.Slice(r.Addrs, func(i, j int) bool {
+			return r.Addrs[i].Expiry < r.Addrs[j].Expiry
+		})
+	}
+
+	// since addresses are sorted by expiration, we find the first survivor and split the
+	// slice on its index.
+	pivot := -1
+	for i, addr := range r.Addrs {
```

Comment:
Nit: `sort.Search`.

(really, the datastore operations will dwarf this... but still).

Reply:
I knew there had to be a simpler way <3

Reply:
I'm not sure if a binary search is what we want here though. I think a linear search works better because items are sorted by expiration in ascending order, and GC is running every 5 minutes (on the lookahead window), so it's likely that the pivot will be near the beginning. But it's just conjecture.

Reply:
I think the binary search will be better. Let's say we have 1000 nodes in the store (we will have more than that). Binary search needs only 10 operations to find the split point. If less than 1% of entries are removed, then a linear search will be better.

In the case of 5000 entries, the break-even point is 0.26% or 13 entries per 5 minutes.

Reply:
Note: we're not talking _nodes_ but _addresses per peer_, which are already sorted by expiration. `clean()` is called on access, modification and during GC, so it's called frequently, which brings the pivot point pretty close to the front.

Most nodes will have no more than 20 addrs (keeping asidse DHT address abusers), so log2(20) is 5 operations. Really I don't think it'll make much difference, but I do think that a linear traversal is better in terms of mechanical sympathy (CPU caches).

Let me know if you feel strongly about this and I'll change it to a binary search ;-)

Reply:
Yeah, let's leave it like it is, if we find it too slow we can think about it.

Reply:
Aah, I didn't read the code before starting above discussion and didn't notice it was a single record, not whole stack from all records.